### PR TITLE
Reland: [Autograd] Use in-place input accumulation fast path for dense Tensors.

### PR DESCRIPTION
**Pull request by robieta**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #90217

Identical to https://github.com/pytorch/pytorch/pull/88339 except with a `.has_storage()` check before `.storage()`.

Differential Revision: [D41737935](https://our.internmc.facebook.com/intern/diff/D41737935/)

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7

### CODE REVIEW

**Review by ngimel — APPROVED** [review]
Hope it sticks this time (which unfortunately is my comment on too many diffs)

File: torch/csrc/autograd/input_buffer.cpp [hunk 1]
```diff
@@ -74,25 +87,38 @@ static void accumulate(
...
+  //  4) The candidate is mutable. Currently only ZeroTensors are immutable.
+  //
+  //  5) The other Tensor is not a Tensor subclass (except sparse), since
+  //     it's hard to predict the semantics of arbitrary subclass behavior.
+
+  if (at::GradMode::is_enabled()) {
+    buffer[pos] = old_var + var;
+  } else if (
+      // ATen doesn't route sparse additions correctly...
+      old_var.is_sparse() || old_var.is_sparse_csr()) {
```
Comment (albanD) [inline torch/csrc/autograd/input_buffer.cpp:111]: cc @cpuhrsch what is the right way (and future proof way) to check if a Tensor is any sparse format?

Comment (cpuhrsch) [reply torch/csrc/autograd/input_buffer.cpp:111]: @albanD - a future proof catch-all is work in progress. Right now it's best to access the layout and compare to the various sparse formats.

`self.layout() == kSparseBsc`, `self.layout() == kSparseBsr`, `self.layout() == kSparseCsc`, `self.layout() == kSparseCsr`, `self.layout() == kSparse` (see [Layout.h](https://github.com/pytorch/pytorch/blob/master/c10/core/Layout.h#L11-L16)).

cc @amjames for visibility. I created a new entry in the roadmap for this so we can track it.

**Pull request closed without merging**